xen/arch/x86/tpm.c

```diff
@@ -985,4 +978,106 @@ void __stdcall tpm_extend_mbi(uint32_t *mbi)
+
+    slrt = __va(os_mle->slrt);
+    if ( slrt->magic != SLR_TABLE_MAGIC )
+        panic("SLRT has invalid magic value: %#08x!\n", slrt->magic);
+    /* XXX: are newer revisions allowed? */
+    if ( slrt->revision != SLR_TABLE_REVISION )
+        panic("SLRT is of unsupported revision: %#04x!\n", slrt->revision);
+    if ( slrt->architecture != SLR_INTEL_TXT )
+        panic("SLRT is for unexpected architecture: %#04x!\n",
+                  slrt->architecture);
+    if ( slrt->size > slrt->max_size )
+        panic("SLRT is larger than its max size: %#08x > %#08x!\n",
+                  slrt->size, slrt->max_size);
+    if ( slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_INTEL_INFO) == NULL )
+        panic("SLRT is missing Intel-specific information!\n");
+
+    policy = (struct slr_entry_policy *)
+        slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_DRTM_POLICY);
+    if (policy == NULL)
+        panic("SLRT is missing DRTM policy!\n");
+
+    /* SLRT policy entry must be the first one (so measuring order matches
+     * policy order) and point to SLRT. */
+    /* XXX: are newer revisions allowed? */
+    if ( policy->revision != SLR_POLICY_REVISION )
+        panic("DRTM policy in SLRT is of unsupported revision: %#04x!\n",
+                  slrt->revision);
+    if ( policy->nr_entries == 0 )
+        panic("DRTM policy in SLRT is empty!\n");
+
+    return policy;
+}
+
+void tpm_take_measurements(void)
+{
+    struct txt_os_mle_data *os_mle;
+    struct slr_table *slrt;
+    struct slr_entry_policy *policy;
+    struct slr_policy_entry *policy_entry;
+    uint16_t i;
+
+    os_mle = txt_os_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
+    slrt = __va(os_mle->slrt);
+
+    policy = tpm_get_policy();
+    policy_entry = (struct slr_policy_entry *)
+        ((uint8_t *)policy + sizeof(*policy));
+    if ( policy_entry[0].entity_type != SLR_ET_MULTIBOOT_INFO ) {
+        panic("First entry of DRTM policy in SLRT has wrong type: %#04x!\n",
+                  policy_entry[0].entity_type);
+    }
+
+    for ( i = 1; i < policy->nr_entries; i++ ) {
+        uint64_t start = policy_entry[i].entity;
+        uint64_t size = policy_entry[i].size;
+        uint32_t event_type = TXT_EVTYPE_UNKNOWN;
+
+        switch ( policy_entry[i].entity_type ) {
+        case SLR_ET_MULTIBOOT_INFO:
+            panic("Duplicated MBI entry in DRTM of Secure Launch at %d\n", i);
+
+        case SLR_ET_SLRT:
+            event_type = TXT_EVTYPE_SLRT;
+            break;
+        case SLR_ET_CMDLINE:
+            event_type = TXT_EVTYPE_KERNEL_CMDLINE;
+            break;
+        case SLR_ET_UEFI_MEMMAP:
+            event_type = TXT_EVTYPE_UEFI_MMAP;
+            break;
+        case SLR_ET_RAMDISK:
+            event_type = TXT_EVTYPE_INITRD;
+            break;
+        case SLR_ET_TXT_OS2MLE:
+            event_type = TXT_EVTYPE_OS2MLE;
+            break;
+        case SLR_ET_MULTIBOOT_MODULE:
```

Comment:
You should drop https://github.com/TrenchBoot/xen/commit/7962cb8b862beabd9de2c78f9dae5b02530e7f07, otherwise kernel and initramfs would be measured twice.

It would be good to check if modules parsed by Xen are the same as here (location and size). At the point where `tpm_take_measurements()` is called modules are not relocated yet, and their addresses are still in MB2 format, so it should be relatively easy to do. With that done it should be safe to drop "Excessive multiboot modules for slaunch - limiting to 2" logic from setup.c, assuming no unmeasured modules are left in MBI.

Reply:
> You should drop [7962cb8](https://github.com/TrenchBoot/xen/commit/7962cb8b862beabd9de2c78f9dae5b02530e7f07), otherwise kernel and initramfs would be measured twice.

Reverted in cd2f2962e88b206f8d289424d2f1b1335f3a7d04 and 2376e6160348fa439e77918e9f976671ad63e3b9.

> It would be good to check if modules parsed by Xen are the same as here (location and size).

https://github.com/TrenchBoot/xen/compare/39a1478db53029b133c75d4f15a0f5774ad92f8a..bab60f90c52351add03aec69881f03a9c30cda6a#diff-5861d423276bd98c31fb2a501569110297cbe6488842e0359435a3af795c24ae